<a href="https://colab.research.google.com/github/romenlaw/NaiveNeuralNetwork/blob/main/tokenisation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

* https://tiktokenizer.vercel.app/
* [MEGABYTE: Predicting Million-byte Sequences with Multiscale Transformers](https://arxiv.org/abs/2305.07185)
* https://en.wikipedia.org/wiki/Byte_pair_encoding

In [1]:
s="""Hello world! 世界这么大，我想去看看。看一看，瞧一瞧了啊！我瞧瞧！
👋🌍
121 + 458 = 679
1215 - 121 = 1094
-1215 + 121 = -1094

Some python code here:
# List of numbers
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Loop through each number in the list
for number in numbers:
    # Check if the number is even
    if number % 2 == 0:
        print(f"{number} is even")
    # Check if the number is odd
    else:
        print(f"{number} is odd")
"""
[list(x.encode("utf-8")) for x in s[:50]]

[[72],
 [101],
 [108],
 [108],
 [111],
 [32],
 [119],
 [111],
 [114],
 [108],
 [100],
 [33],
 [32],
 [228, 184, 150],
 [231, 149, 140],
 [232, 191, 153],
 [228, 185, 136],
 [229, 164, 167],
 [239, 188, 140],
 [230, 136, 145],
 [230, 131, 179],
 [229, 142, 187],
 [231, 156, 139],
 [231, 156, 139],
 [227, 128, 130],
 [231, 156, 139],
 [228, 184, 128],
 [231, 156, 139],
 [239, 188, 140],
 [231, 158, 167],
 [228, 184, 128],
 [231, 158, 167],
 [228, 186, 134],
 [229, 149, 138],
 [239, 188, 129],
 [230, 136, 145],
 [231, 158, 167],
 [231, 158, 167],
 [239, 188, 129],
 [10],
 [240, 159, 145, 139],
 [240, 159, 140, 141],
 [10],
 [49],
 [50],
 [49],
 [32],
 [43],
 [32],
 [52]]

In [2]:
[ord(x) for x in s[:50]]

[72,
 101,
 108,
 108,
 111,
 32,
 119,
 111,
 114,
 108,
 100,
 33,
 32,
 19990,
 30028,
 36825,
 20040,
 22823,
 65292,
 25105,
 24819,
 21435,
 30475,
 30475,
 12290,
 30475,
 19968,
 30475,
 65292,
 30631,
 19968,
 30631,
 20102,
 21834,
 65281,
 25105,
 30631,
 30631,
 65281,
 10,
 128075,
 127757,
 10,
 49,
 50,
 49,
 32,
 43,
 32,
 52]

In [3]:
len('世世界')

3

# Byte Pair encoding

In [4]:
raw_bytes = [x for x in s.encode('utf-8')]
print(s, len(s), len(raw_bytes), raw_bytes)

Hello world! 世界这么大，我想去看看。看一看，瞧一瞧了啊！我瞧瞧！
👋🌍
121 + 458 = 679
1215 - 121 = 1094
-1215 + 121 = -1094

Some python code here:
# List of numbers
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Loop through each number in the list
for number in numbers:
    # Check if the number is even
    if number % 2 == 0:
        print(f"{number} is even")
    # Check if the number is odd
    else:
        print(f"{number} is odd")
 414 472 [72, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100, 33, 32, 228, 184, 150, 231, 149, 140, 232, 191, 153, 228, 185, 136, 229, 164, 167, 239, 188, 140, 230, 136, 145, 230, 131, 179, 229, 142, 187, 231, 156, 139, 231, 156, 139, 227, 128, 130, 231, 156, 139, 228, 184, 128, 231, 156, 139, 239, 188, 140, 231, 158, 167, 228, 184, 128, 231, 158, 167, 228, 186, 134, 229, 149, 138, 239, 188, 129, 230, 136, 145, 231, 158, 167, 231, 158, 167, 239, 188, 129, 10, 240, 159, 145, 139, 240, 159, 140, 141, 10, 49, 50, 49, 32, 43, 32, 52, 53, 56, 32, 61, 32, 54, 55, 57, 10, 49, 50, 49, 5

In [5]:
def count_bytepairs(bytes):
  counts = {}
  for pair in zip(bytes, bytes[1:]):
    counts[pair] = counts.get(pair, 0) + 1
  counts = dict(sorted(counts.items(), key=lambda item: item[1], reverse=True))
  return counts

In [6]:
counts = count_bytepairs(raw_bytes)

In [7]:
len(counts), list(counts)[:5]

(224, [(32, 32), (101, 114), (110, 117), (117, 109), (109, 98)])

In [8]:
counts

{(32, 32): 26,
 (101, 114): 11,
 (110, 117): 10,
 (117, 109): 10,
 (109, 98): 10,
 (98, 101): 10,
 (44, 32): 9,
 (32, 105): 9,
 (32, 110): 7,
 (104, 101): 6,
 (105, 115): 6,
 (114, 32): 6,
 (10, 32): 6,
 (49, 50): 5,
 (50, 49): 5,
 (32, 61): 5,
 (61, 32): 5,
 (101, 32): 5,
 (116, 104): 5,
 (115, 32): 5,
 (239, 188): 4,
 (231, 156): 4,
 (156, 139): 4,
 (231, 158): 4,
 (158, 167): 4,
 (32, 49): 4,
 (58, 10): 4,
 (35, 32): 4,
 (102, 32): 4,
 (32, 116): 4,
 (32, 101): 4,
 (105, 110): 4,
 (228, 184): 3,
 (49, 32): 3,
 (49, 48): 3,
 (32, 112): 3,
 (110, 32): 3,
 (111, 100): 3,
 (32, 111): 3,
 (114, 115): 3,
 (105, 102): 3,
 (101, 108): 2,
 (111, 114): 2,
 (167, 239): 2,
 (188, 140): 2,
 (230, 136): 2,
 (136, 145): 2,
 (184, 128): 2,
 (128, 231): 2,
 (167, 228): 2,
 (188, 129): 2,
 (240, 159): 2,
 (10, 49): 2,
 (32, 43): 2,
 (43, 32): 2,
 (32, 52): 2,
 (32, 54): 2,
 (49, 53): 2,
 (53, 32): 2,
 (32, 45): 2,
 (48, 57): 2,
 (57, 52): 2,
 (52, 10): 2,
 (45, 49): 2,
 (10, 10): 2,
 (101, 58): 2,
 (

In [9]:
# utf-8 stops at 255, so our new tokens will start at 256

def merge_pair(raw_bytes, pair):
  """pair is a list containing two integers, representing the utf-8 code of
  the 2 characters to be replaced
  """
  assert isinstance(pair, list) and len(pair)==2, "pair must be a list of two integers"
  max_val = max(raw_bytes)
  NEW_TOKEN = [256 if max_val<256 else max_val+1]
  for i in range(len(raw_bytes)-1, 0, -1):  # loop backwards
    #print(i-1, i)
    if raw_bytes[i-1:i+1] == pair:
      raw_bytes[i-1:i+1] = NEW_TOKEN

  return raw_bytes


In [10]:
out = merge_pair(raw_bytes, [32,32])
len(out)

456

In [11]:
counts = count_bytepairs(raw_bytes)
key, value = next(iter(counts.items())) # get first entry of dict
while value>1:
  for pair, count in counts.items():
    if count>1:
      merge_pair(raw_bytes, list(pair))
    else:
      break
  counts = count_bytepairs(raw_bytes)
  key, value = next(iter(counts.items())) # get first entry of dict

In [12]:
print(len(raw_bytes), max(raw_bytes), raw_bytes)

190 358 [72, 288, 108, 111, 32, 119, 289, 108, 100, 33, 32, 281, 150, 231, 149, 140, 232, 191, 153, 228, 185, 136, 229, 164, 320, 140, 322, 230, 131, 179, 229, 142, 187, 312, 312, 227, 128, 130, 312, 323, 312, 321, 315, 323, 274, 291, 186, 134, 229, 149, 138, 324, 322, 315, 315, 324, 10, 292, 145, 139, 292, 140, 141, 325, 282, 294, 52, 53, 56, 267, 295, 55, 57, 325, 296, 297, 346, 275, 298, 300, 45, 326, 293, 346, 297, 283, 299, 301, 83, 111, 109, 269, 112, 121, 270, 111, 284, 99, 285, 269, 104, 257, 329, 330, 331, 286, 278, 345, 10, 345, 267, 32, 91, 49, 261, 50, 261, 51, 261, 52, 261, 53, 261, 54, 261, 55, 261, 56, 261, 57, 261, 283, 93, 301, 330, 111, 111, 112, 32, 270, 114, 111, 117, 103, 104, 287, 97, 99, 303, 344, 110, 279, 317, 108, 331, 10, 102, 111, 265, 344, 284, 345, 318, 357, 335, 306, 319, 105, 347, 257, 32, 37, 302, 267, 268, 48, 358, 336, 310, 319, 357, 342, 100, 319, 288, 115, 101, 358, 343, 341]


# Byte Pair Implementation

In [38]:
s = """Hello world! 世界这么大，我想去看看。看一看，瞧一瞧了啊！我瞧瞧！
👋🌍
121 + 458 = 679
1215 - 121 = 1094
-1215 + 121 = -1094
Python Lists

bookmark_border
Python has a great built-in list type named "list". List literals are written within square brackets [ ]. Lists work similarly to strings -- use the len() function and square brackets [ ] to access data, with the first element at index 0. (See the official python.org list docs.)


  colors = ['red', 'blue', 'green']
  print(colors[0])    ## red
  print(colors[2])    ## green
  print(len(colors))  ## 3
list of strings 'red' 'blue 'green'

Assignment with an = on lists does not make a copy. Instead, assignment makes the two variables point to the one list in memory.


  b = colors   ## Does not copy the list
both colors and b point to the one list

The "empty list" is just an empty pair of brackets [ ]. The '+' works to append two lists, so [1, 2] + [3, 4] yields [1, 2, 3, 4] (this is just like + with strings).

FOR and IN
Python's *for* and *in* constructs are extremely useful, and the first use of them we'll see is with lists. The *for* construct -- for var in list -- is an easy way to look at each element in a list (or other collection). Do not add or remove from the list during iteration.


  squares = [1, 4, 9, 16]
  sum = 0
  for num in squares:
    sum += num
  print(sum)  ## 30
If you know what sort of thing is in the list, use a variable name in the loop that captures that information such as "num", or "name", or "url". Since Python code does not have other syntax to remind you of types, your variable names are a key way for you to keep straight what is going on. (This is a little misleading. As you gain more exposure to python, you'll see references to type hints which allow you to add typing information to your function definitions. Python doesn't use these type hints when it runs your programs. They are used by other programs such as IDEs (integrated development environments) and static analysis tools like linters/type checkers to validate if your functions are called with compatible arguments.)

The *in* construct on its own is an easy way to test if an element appears in a list (or other collection) -- value in collection -- tests if the value is in the collection, returning True/False.


  list = ['larry', 'curly', 'moe']
  if 'curly' in list:
    print('yay') ## yay
The for/in constructs are very commonly used in Python code and work on data types other than list, so you should just memorize their syntax. You may have habits from other languages where you start manually iterating over a collection, where in Python you should just use for/in.

You can also use for/in to work on a string. The string acts like a list of its chars, so for ch in s: print(ch) prints all the chars in a string.

Range
The range(n) function yields the numbers 0, 1, ... n-1, and range(a, b) returns a, a+1, ... b-1 -- up to but not including the last number. The combination of the for-loop and the range() function allow you to build a traditional numeric for loop:


  ## print the numbers from 0 through 99
  for i in range(100):
    print(i)
There is a variant xrange() which avoids the cost of building the whole list for performance sensitive cases (in Python 3, range() will have the good performance behavior and you can forget about xrange()).

While Loop
Python also has the standard while-loop, and the *break* and *continue* statements work as in C++ and Java, altering the course of the innermost loop. The above for/in loops solves the common case of iterating over every element in a list, but the while loop gives you total control over the index numbers. Here's a while loop which accesses every 3rd element in a list:


  ## Access every 3rd element in a list
  i = 0
  while i < len(a):
    print(a[i])
    i = i + 3
List Methods
Here are some other common list methods.

list.append(elem) -- adds a single element to the end of the list. Common error: does not return the new list, just modifies the original.
list.insert(index, elem) -- inserts the element at the given index, shifting elements to the right.
list.extend(list2) adds the elements in list2 to the end of the list. Using + or += on a list is similar to using extend().
list.index(elem) -- searches for the given element from the start of the list and returns its index. Throws a ValueError if the element does not appear (use "in" to check without a ValueError).
list.remove(elem) -- searches for the first instance of the given element and removes it (throws ValueError if not present)
list.sort() -- sorts the list in place (does not return it). (The sorted() function shown later is preferred.)
list.reverse() -- reverses the list in place (does not return it)
list.pop(index) -- removes and returns the element at the given index. Returns the rightmost element if index is omitted (roughly the opposite of append()).
Notice that these are *methods* on a list object, while len() is a function that takes the list (or string or whatever) as an argument.


  list = ['larry', 'curly', 'moe']
  list.append('shemp')         ## append elem at end
  list.insert(0, 'xxx')        ## insert elem at index 0
  list.extend(['yyy', 'zzz'])  ## add list of elems at end
  print(list)  ## ['xxx', 'larry', 'curly', 'moe', 'shemp', 'yyy', 'zzz']
  print(list.index('curly'))    ## 2

  list.remove('curly')         ## search and remove that element
  list.pop(1)                  ## removes and returns 'larry'
  print(list)  ## ['xxx', 'moe', 'shemp', 'yyy', 'zzz']
Common error: note that the above methods do not *return* the modified list, they just modify the original list.


  list = [1, 2, 3]
  print(list.append(4))   ## NO, does not work, append() returns None
  ## Correct pattern:
  list.append(4)
  print(list)  ## [1, 2, 3, 4]
List Build Up
One common pattern is to start a list as the empty list [], then use append() or extend() to add elements to it:


  list = []          ## Start as the empty list
  list.append('a')   ## Use append() to add elements
  list.append('b')
List Slices
Slices work on lists just as with strings, and can also be used to change sub-parts of the list.


  list = ['a', 'b', 'c', 'd']
  print(list[1:-1])   ## ['b', 'c']
  list[0:2] = 'z'    ## replace ['a', 'b'] with ['z']
  print(list)         ## ['z', 'c', 'd']
Exercise: list1.py
To practice the material in this section, try the problems in list1.py that do not use sorting (in the Basic Exercises).

Was this helpful?

Except as otherwise noted, the content of this page is licensed under the Creative Commons Attribution 4.0 License, and code samples are licensed under the Apache 2.0 License. For details, see the Google Developers Site Policies. Java is a registered trademark of Oracle and/or its affiliates.

Last updated 2024-07-23 UTC.
"""
tokens = s.encode("utf-8")

In [60]:
def count_bytepairs(bytes):
  counts = {}
  for pair in zip(bytes, bytes[1:]):
    counts[pair] = counts.get(pair, 0) + 1
  counts = dict(sorted(counts.items(), key=lambda item: item[1], reverse=True))
  return counts

def merge_bytepair(raw_bytes, pair, new_token):
  """pair is a list containing two integers, representing the utf-8 code of
  the 2 characters to be replaced
  """
  assert isinstance(pair, list) and len(pair)==2, "pair must be a list of two integers"
  for i in range(len(raw_bytes)-1, 0, -1):  # loop backwards
    #print(i-1, i)
    if raw_bytes[i-1:i+1] == pair:
      raw_bytes[i-1:i+1] = [new_token]

  return raw_bytes

# hyper parameters
vocab_size = 296
num_merges = vocab_size - 256
merges = {} # build up a lookup of what has been merged into what
vocab = { i:bytes([i]) for i in range(256)}

def encode(s):
  """Given a string, return a list of integers using utf-8 and byte pair
  """
  assert isinstance(s, str), f"expect string, but got {type(s).__name__}"
  tokens = s.encode("utf-8")
  token_copy = list(tokens)
  counts = count_bytepairs(token_copy)
  if len(counts) <=1: # otherwise next(counts_iter) in below loop will fail
      return token_copy

  new_token = 255
  counts_iter = iter(counts.items())
  #for i in range(num_merges):
  while True:
    new_token += 1
    pair, count = next(counts_iter)
    if count>1:
      pair_copy = list(pair)
      merge_bytepair(token_copy, pair_copy, new_token)
      merges[pair] = new_token
    else:
      break

  vocab.update({tok:bytes([b1])+bytes([b2]) for (b1,b2), tok in merges.items()})
  return token_copy

In [61]:
enc = encode(s)
print(len(enc), max(enc), enc[:100])

3959 742 [72, 302, 108, 292, 119, 273, 437, 33, 32, 588, 150, 231, 149, 140, 232, 191, 153, 228, 185, 136, 229, 164, 167, 536, 140, 634, 145, 230, 131, 179, 229, 142, 187, 537, 139, 537, 139, 227, 128, 130, 537, 139, 588, 128, 537, 139, 536, 140, 539, 167, 588, 128, 539, 638, 186, 134, 229, 149, 138, 536, 129, 634, 145, 539, 167, 539, 167, 536, 129, 10, 640, 145, 139, 640, 140, 141, 10, 511, 49, 438, 478, 53, 56, 32, 349, 54, 55, 642, 511, 643, 364, 32, 511, 541, 349, 589, 646, 10, 45, 511]


In [41]:
len(tokens), max(tokens), list(tokens)[:10]

(6843, 240, [72, 101, 108, 108, 111, 32, 119, 111, 114, 108])

In [42]:
print(f"compression ratio: {len(tokens) / len(enc):.2f}X", merges)

compression ratio: 1.73X {(101, 32): 256, (32, 32): 257, (115, 32): 258, (32, 116): 259, (116, 104): 260, (115, 116): 261, (116, 32): 262, (32, 97): 263, (105, 110): 264, (105, 115): 265, (104, 101): 266, (108, 105): 267, (110, 32): 268, (32, 105): 269, (32, 108): 270, (44, 32): 271, (101, 110): 272, (111, 114): 273, (114, 101): 274, (111, 110): 275, (101, 114): 276, (110, 100): 277, (110, 116): 278, (108, 101): 279, (32, 111): 280, (32, 115): 281, (100, 32): 282, (97, 110): 283, (101, 109): 284, (101, 115): 285, (114, 32): 286, (115, 101): 287, (41, 32): 288, (32, 101): 289, (10, 32): 290, (97, 114): 291, (111, 32): 292, (114, 105): 293, (32, 99): 294, (32, 119): 295, (97, 116): 296, (109, 101): 297, (32, 112): 298, (110, 103): 299, (10, 10): 300, (116, 101): 301, (101, 108): 302, (32, 39): 303, (118, 101): 304, (116, 115): 305, (105, 116): 306, (116, 105): 307, (99, 111): 308, (116, 111): 309, (32, 102): 310, (32, 114): 311, (97, 32): 312, (32, 110): 313, (121, 32): 314, (102, 32): 3

In [43]:
max(merges, key=merges.get)

(46, 48)

# Decoding
Given a list of tokens (integers), return python string.

In [62]:
print(vocab)

{0: b'\x00', 1: b'\x01', 2: b'\x02', 3: b'\x03', 4: b'\x04', 5: b'\x05', 6: b'\x06', 7: b'\x07', 8: b'\x08', 9: b'\t', 10: b'\n', 11: b'\x0b', 12: b'\x0c', 13: b'\r', 14: b'\x0e', 15: b'\x0f', 16: b'\x10', 17: b'\x11', 18: b'\x12', 19: b'\x13', 20: b'\x14', 21: b'\x15', 22: b'\x16', 23: b'\x17', 24: b'\x18', 25: b'\x19', 26: b'\x1a', 27: b'\x1b', 28: b'\x1c', 29: b'\x1d', 30: b'\x1e', 31: b'\x1f', 32: b' ', 33: b'!', 34: b'"', 35: b'#', 36: b'$', 37: b'%', 38: b'&', 39: b"'", 40: b'(', 41: b')', 42: b'*', 43: b'+', 44: b',', 45: b'-', 46: b'.', 47: b'/', 48: b'0', 49: b'1', 50: b'2', 51: b'3', 52: b'4', 53: b'5', 54: b'6', 55: b'7', 56: b'8', 57: b'9', 58: b':', 59: b';', 60: b'<', 61: b'=', 62: b'>', 63: b'?', 64: b'@', 65: b'A', 66: b'B', 67: b'C', 68: b'D', 69: b'E', 70: b'F', 71: b'G', 72: b'H', 73: b'I', 74: b'J', 75: b'K', 76: b'L', 77: b'M', 78: b'N', 79: b'O', 80: b'P', 81: b'Q', 82: b'R', 83: b'S', 84: b'T', 85: b'U', 86: b'V', 87: b'W', 88: b'X', 89: b'Y', 90: b'Z', 91: b'[',

In [46]:
def decode(ids):
  """Given a list of tokens (integers), return a python string
  """
  assert isinstance(ids, list), "ids must be list"
  assert all((isinstance(i, int) and i>=0 ) for i in ids), \
    "ids must be a list of integers in [0, vocab_size)"
#   out = bytes()
#   for b in ids:
#     out += vocab[b]
  out = b"".join(vocab[b] for b in ids)
  return out.decode("utf-8", errors='replace') # see utf-8 spec

In [22]:
decode([66,32,67]), decode([128])

('B C', '�')

In [23]:
s="""assert isinstance(ids, list), "ids must be a list" """
tokens = s.encode("utf-8")
decode(list(tokens))

'assert isinstance(ids, list), "ids must be a list" '

In [63]:
decode(enc)==s, decode(enc)

(True,
 'Hello world! 世界这么大，我想去看看。看一看，瞧一瞧了啊！我瞧瞧！\n👋🌍\n121 + 458 = 679\n1215 - 121 = 1094\n-1215 + 121 = -1094\nPython Lists\n\nbookmark_border\nPython has a great built-in list type named "list". List literals are written within square brackets [ ]. Lists work similarly to strings -- use the len() function and square brackets [ ] to access data, with the first element at index 0. (See the official python.org list docs.)\n\n\n  colors = [\'red\', \'blue\', \'green\']\n  print(colors[0])    ## red\n  print(colors[2])    ## green\n  print(len(colors))  ## 3\nlist of strings \'red\' \'blue \'green\'\n\nAssignment with an = on lists does not make a copy. Instead, assignment makes the two variables point to the one list in memory.\n\n\n  b = colors   ## Does not copy the list\nboth colors and b point to the one list\n\nThe "empty list" is just an empty pair of brackets [ ]. The \'+\' works to append two lists, so [1, 2] + [3, 4] yields [1, 2, 3, 4] (this is just like + with strings).\n\nFOR 

In [50]:
encode("")

[]

In [51]:
len(counts)

189

# forcing split (gpt2)
https://github.com/openai/gpt-2/blob/master/src/encoder.py

In [27]:
import regex as re
gpt2pat = re.compile(r"""'s|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+""")
re.findall(gpt2pat, "Hello world!!!      How've you been??         ")

['Hello',
 ' world',
 '!!!',
 '     ',
 ' How',
 "'ve",
 ' you',
 ' been',
 '??',
 '         ']

# tiktokeniser
* https://github.com/openai/tiktoken
* https://github.com/openai/tiktoken/blob/main/tiktoken_ext/openai_public.py

In [52]:
!pip install tiktoken
import tiktoken

In [53]:
enc = tiktoken.get_encoding("gpt2")
print(enc.encode("    Hello world!!!   How've you been??  "))

enc = tiktoken.get_encoding("cl100k_base") # used by gpt4
print(enc.encode("    Hello world!!!   How've you been??  "))

enc = enc = tiktoken.encoding_for_model("gpt-4o")
print(enc.encode("    Hello world!!!   How've you been??  "))

[220, 220, 220, 18435, 995, 10185, 220, 220, 1374, 1053, 345, 587, 3548, 220, 220]
[262, 22691, 1917, 12340, 256, 2650, 3077, 499, 1027, 7801, 256]
[271, 32949, 2375, 10880, 256, 3253, 7341, 481, 1339, 6961, 256]


In [30]:
gpt4pat=r"""'(?i:[sdmt]|ll|ve|re)|[^\r\n\p{L}\p{N}]?+\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]++[\r\n]*|\s*[\r\n]|\s+(?!\S)|\s+"""
re.findall(gpt4pat, "Hello world!!!      How've you been??  123  56788900       ")

['Hello',
 ' world',
 '!!!',
 '     ',
 ' How',
 "'ve",
 ' you',
 ' been',
 '??',
 ' ',
 ' ',
 '123',
 ' ',
 ' ',
 '567',
 '889',
 '00',
 '       ']

In [64]:
!wget https://openaipublic.blob.core.windows.net/gpt-2/encodings/main/vocab.bpe
!wget https://openaipublic.blob.core.windows.net/gpt-2/encodings/main/encoder.json

--2024-09-18 09:44:53--  https://openaipublic.blob.core.windows.net/gpt-2/encodings/main/vocab.bpe
Resolving openaipublic.blob.core.windows.net (openaipublic.blob.core.windows.net)... 57.150.97.129
Connecting to openaipublic.blob.core.windows.net (openaipublic.blob.core.windows.net)|57.150.97.129|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 456318 (446K) [application/octet-stream]
Saving to: ‘vocab.bpe’

vocab.bpe           100%[===================>] 445.62K  1.31MB/s    in 0.3s    

2024-09-18 09:44:53 (1.31 MB/s) - ‘vocab.bpe’ saved [456318/456318]

--2024-09-18 09:44:53--  https://openaipublic.blob.core.windows.net/gpt-2/encodings/main/encoder.json
Resolving openaipublic.blob.core.windows.net (openaipublic.blob.core.windows.net)... 57.150.97.129
Connecting to openaipublic.blob.core.windows.net (openaipublic.blob.core.windows.net)|57.150.97.129|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1042301 (1018K) [application/json]
Sav

In [65]:
import os, json

with open("encoder.json", "r") as f:
  encoder = json.load(f)

with open("vocab.bpe", "r") as f:
  bpe_data = f.read()
# ignore first line (version...) and last (new line)
bpe_merges = [tuple(merge_str.split()) for merge_str in bpe_data.split('\n')[1:-1]]

In [75]:
# this is similar to our vocab, but key and values are swapped
len(encoder), dict(list(encoder.items())[:10])

(50257,
 {'!': 0,
  '"': 1,
  '#': 2,
  '$': 3,
  '%': 4,
  '&': 5,
  "'": 6,
  '(': 7,
  ')': 8,
  '*': 9})

In [76]:
# this is similar to our merges
len(bpe_merges), bpe_merges[:10]

(50000,
 [('Ġ', 't'),
  ('Ġ', 'a'),
  ('h', 'e'),
  ('i', 'n'),
  ('r', 'e'),
  ('o', 'n'),
  ('Ġt', 'he'),
  ('e', 'r'),
  ('Ġ', 's'),
  ('a', 't')])

# special tokens
GPT has 256 raw bytes, and 50000 merges, and 1 special token, so totally 50257 tokens.

In [78]:
len(encoder), len(bpe_merges)

(50257, 50000)

In [81]:
encoder['<|endoftext|>']

50256

In [82]:
dict(list(encoder.items())[50253:])

{'ĠCollider': 50253,
 'Ġinformants': 50254,
 'Ġgazed': 50255,
 '<|endoftext|>': 50256}

# sentencepiece
* https://github.com/google/sentencepiece/tree/d8f741853847553169444afc12c00f4bbff3e9ce

*   https://github.com/google/sentencepiece/blob/d8f741853847553169444afc12c00f4bbff3e9ce/doc/options.md
* https://github.com/google/sentencepiece/blob/d8f741853847553169444afc12c00f4bbff3e9ce/src/sentencepiece_model.proto



In [84]:
!wget "https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt"
import sentencepiece as spm

--2024-09-18 10:20:57--  https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1115394 (1.1M) [text/plain]
Saving to: ‘input.txt’

input.txt           100%[===================>]   1.06M  --.-KB/s    in 0.08s   

2024-09-18 10:20:57 (12.6 MB/s) - ‘input.txt’ saved [1115394/1115394]



In [87]:
# train a sentencepiece model on it
# the settings here are (best effort) those used for training Llama 2
import os

options = dict(
    # input spec
    input="input.txt",
    input_format="text",
    # output spec
    model_prefix="tok400", # output file prefix
    # algorithm specific
    # BPE algo
    model_type="bpe",
    vocab_size=400,
    # normalisation
    normalization_rule_name="identity", # turn off normalisation
    remove_extra_whitespaces=False,
    input_sentence_size=200000000, # max number of training sentences
    max_sentence_length=4192, # max number of bytes per sentence
    seed_sentencepiece_size=1000000,
    shuffle_input_sentence=True,
    # rare word treatment
    character_coverage=0.99995, # if character is very rare, it will be igored (not put in vocab)
    byte_fallback=True,
    # merge rules
    split_digits=True,
    split_by_unicode_script=True,
    split_by_whitespace=True,
    split_by_number=True,
    max_sentencepiece_length=16,
    add_dummy_prefix=True,
    allow_whitespace_only_pieces=True,
    # special tokens
    unk_id=0, # UNK token is mandatory
    bos_id=1, # others are optional, if not needed, set to -1
    eos_id=2,
    pad_id=-1, # -1 means don't use it
    # systems
    num_threads=os.cpu_count(),
)
spm.SentencePieceTrainer.train(**options)

In [88]:
sp = spm.SentencePieceProcessor()
sp.load("tok400.model")
vocab = [[sp.IdToPiece(idx), idx] for idx in range(sp.get_piece_size())]
vocab

[['<unk>', 0],
 ['<s>', 1],
 ['</s>', 2],
 ['<0x00>', 3],
 ['<0x01>', 4],
 ['<0x02>', 5],
 ['<0x03>', 6],
 ['<0x04>', 7],
 ['<0x05>', 8],
 ['<0x06>', 9],
 ['<0x07>', 10],
 ['<0x08>', 11],
 ['<0x09>', 12],
 ['<0x0A>', 13],
 ['<0x0B>', 14],
 ['<0x0C>', 15],
 ['<0x0D>', 16],
 ['<0x0E>', 17],
 ['<0x0F>', 18],
 ['<0x10>', 19],
 ['<0x11>', 20],
 ['<0x12>', 21],
 ['<0x13>', 22],
 ['<0x14>', 23],
 ['<0x15>', 24],
 ['<0x16>', 25],
 ['<0x17>', 26],
 ['<0x18>', 27],
 ['<0x19>', 28],
 ['<0x1A>', 29],
 ['<0x1B>', 30],
 ['<0x1C>', 31],
 ['<0x1D>', 32],
 ['<0x1E>', 33],
 ['<0x1F>', 34],
 ['<0x20>', 35],
 ['<0x21>', 36],
 ['<0x22>', 37],
 ['<0x23>', 38],
 ['<0x24>', 39],
 ['<0x25>', 40],
 ['<0x26>', 41],
 ['<0x27>', 42],
 ['<0x28>', 43],
 ['<0x29>', 44],
 ['<0x2A>', 45],
 ['<0x2B>', 46],
 ['<0x2C>', 47],
 ['<0x2D>', 48],
 ['<0x2E>', 49],
 ['<0x2F>', 50],
 ['<0x30>', 51],
 ['<0x31>', 52],
 ['<0x32>', 53],
 ['<0x33>', 54],
 ['<0x34>', 55],
 ['<0x35>', 56],
 ['<0x36>', 57],
 ['<0x37>', 58],
 ['<0x38>', 5

In [89]:
ids = sp.Encode("Hello world! 世界这么大")
print(ids)

[329, 340, 277, 342, 265, 273, 318, 384, 339, 231, 187, 153, 234, 152, 143, 235, 194, 156, 231, 188, 139, 232, 167, 170]


the chinese characters are their corresponding 3-byte utf-8 codes.
this is because those chinese characters are not in the training set, they are treated as UNK (i.e. 0); but because byte_fallback=True, they fallback to their utf-8 bytes.

Also, notice the extra spece in front of 'Hello'. This is because of add_dummy_prefix=True.

Read the [protobuf](https://github.com/google/sentencepiece/blob/d8f741853847553169444afc12c00f4bbff3e9ce/src/sentencepiece_model.proto) file for more details.

In [90]:


print([sp.IdToPiece(idx) for idx in ids])

['▁H', 'e', 'll', 'o', '▁w', 'or', 'ld', '!', '▁', '<0xE4>', '<0xB8>', '<0x96>', '<0xE7>', '<0x95>', '<0x8C>', '<0xE8>', '<0xBF>', '<0x99>', '<0xE4>', '<0xB9>', '<0x88>', '<0xE5>', '<0xA4>', '<0xA7>']


[Learning to Compress Prompts with Gist Tokens](https://arxiv.org/abs/2304.08467)

https://platform.openai.com/playground/complete